staging/src/k8s.io/client-go/util/workqueue/delaying_queue_test.go

```diff
@@ -69,20 +322,310 @@ func TestSimpleQueue(t *testing.T) {
+	})
+
+	// step past the original delayed item and process it from the active queue
+	fakeClock.Step(5 * time.Second)
+	if err := waitForAdded(q, 2); err != nil {
+		t.Fatalf("the first item should have been queued")
 	}
-	q.AddAfter(first, 70*time.Millisecond)
-	if err := waitForWaitingQueueToFill(q); err != nil {
+	q.AssertState(t, QueueState{
+		active: []interface{}{first, third},
+	})
+	// pop them both off the queue
+	item, _ := q.Get()
+	q.Done(item)
+	item, _ = q.Get()
+	q.Done(item)
+
+	// add the second item again with DropIfWaiting which should now succeed
+	q.AddWithOptions(second, DelayingOptions{WhenWaiting: TakeExisting})
+	q.WaitForQueueFill()
+	q.AssertState(t, QueueState{
+		active: []interface{}{second},
+	})
+}
+
+// TestDelayedInsertTakeShorter - tests the insert behaviour with this WhenWaiting behaviour (the default)
+func TestDelayedInsertTakeShorter(t *testing.T) {
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	q := NewTestDelayingQueueWithCustomClock(t, fakeClock, "")
+
+	first := "baz"
+	second := "foo"
+	third := "foo"
+
+	// add an item into the waiting queue.
+	q.AddAfter(first, 10*time.Second)
+	q.AddAfter(second, 10*time.Second)
+	// attempt add delayed again with a shorter time time out and TakeShorter option.
+	q.AddWithOptions(third, DelayingOptions{Duration: 1 * time.Second, WhenWaiting: TakeShorter})
+	q.WaitForQueueFill()
+	// step past the shorter item and confirm that it got queued.
+	fakeClock.Step(2 * time.Second)
+	if err := waitForAdded(q, 1); err != nil {
+		t.Fatalf("expected success, got: %v", err)
+	}
+	// we should be left with baz in waiting and foo should now be active
+	q.AssertState(t, QueueState{
+		active:  []interface{}{third},
+		waiting: []interface{}{first},
+	})
+}
+
+// TestDelayedInsertTakeLonger- tests the insert behaviour with this WhenWaiting behaviour.
+func TestDelayedInsertTakeLonger(t *testing.T) {
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	q := NewTestDelayingQueueWithCustomClock(t, fakeClock, "")
+
+	first := "baz"
+	second := "foo"
+	third := "foo"
+	fourth := "foo"
+
+	// add an item into the waiting queue.
+	q.AddAfter(first, 60*time.Second)
+	q.AddAfter(second, 5*time.Second)
+	// attempt add delayed again with a shorter time time out and TakeLonger option.
+	q.AddWithOptions(third, DelayingOptions{Duration: 1 * time.Second, WhenWaiting: TakeLonger})
+	q.WaitForQueueFill()
+	// step past the shorter item and confirm that it didn't get queued
+	fakeClock.Step(2 * time.Second)
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
+		t.Fatalf("expected timeout, got: %v", err)
+	}
+	// Add a longer delay with the expectation of pushing the ready later...
+	q.AddWithOptions(fourth, DelayingOptions{Duration: 20 * time.Second, WhenWaiting: TakeLonger})
+	q.WaitForQueueFill()
+	// step past the original delayed item and check it queues.
+	fakeClock.Step(4 * time.Second)
+	if err := waitForAdded(q, 1); err != wait.ErrWaitTimeout {
```

Comment:
The size of the times are irrelevant as we step the clock forward ourselves.  The delay in the tests (currently 30s) is because we need to wait for an item and time out after waiting 10s in order to prove an item wasn't queued.  This is how the original test suite does it.  This is partly the reason for wanting assertState for a number of my tests because it is much quicker.

Reply:
I think this will depend on how slow the k8s test platforms can be.  On my laptop it could easily be dialled down to a short time, but I wonder why the original tests had this set at 10s?  The tests will become flakey if running on a slow executor with a vert short delay.  Want to try 5s instead?